data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java

```diff
@@ -74,111 +77,106 @@ private TableMigrationUtil() {
+    } catch (UncheckedIOException e) {
+      throw new RuntimeException("Unable to read Avro file: " +
+              path, e);
     }
   }
 
-  private static List<DataFile> listOrcPartition(Map<String, String> partitionPath, String partitionUri,
-                                                 PartitionSpec spec, Configuration conf,
-                                                 MetricsConfig metricsSpec, NameMapping mapping) {
+  private static Metrics getParquetMerics(Path path,  Configuration conf,
+                                          MetricsConfig metricsSpec, NameMapping mapping) {
     try {
-      Path partition = new Path(partitionUri);
-      FileSystem fs = partition.getFileSystem(conf);
+      InputFile file = HadoopInputFile.fromPath(path, conf);
+      return ParquetUtil.fileMetrics(file, metricsSpec, mapping);
+    } catch (UncheckedIOException e) {
+      throw new RuntimeException("Unable to read the metrics of the Parquet file: " +
+              path, e);
+    }
+  }
 
-      return Arrays.stream(fs.listStatus(partition, HIDDEN_PATH_FILTER))
-          .filter(FileStatus::isFile)
-          .map(stat -> {
-            Metrics metrics = OrcMetrics.fromInputFile(HadoopInputFile.fromPath(stat.getPath(), conf),
-                metricsSpec, mapping);
-            String partitionKey = spec.fields().stream()
-                .map(PartitionField::name)
-                .map(name -> String.format("%s=%s", name, partitionPath.get(name)))
-                .collect(Collectors.joining("/"));
+  private static Metrics getOrcMerics(Path path,  Configuration conf,
+                                          MetricsConfig metricsSpec, NameMapping mapping) {
+    try {
+      return OrcMetrics.fromInputFile(HadoopInputFile.fromPath(path, conf),
+              metricsSpec, mapping);
+    } catch (UncheckedIOException e) {
+      throw new RuntimeException("Unable to read the metrics of the Orc file: " +
+              path, e);
+    }
+  }
 
-            return DataFiles.builder(spec)
-                .withPath(stat.getPath().toString())
-                .withFormat("orc")
-                .withFileSizeInBytes(stat.getLen())
-                .withMetrics(metrics)
-                .withPartitionPath(partitionKey)
-                .build();
+  private static DataFile buildDataFile(FileStatus stat, String partitionKey,
+                                      PartitionSpec spec, Metrics metrics, String format) {
+    return  DataFiles.builder(spec)
+            .withPath(stat.getPath().toString())
+            .withFormat(format)
+            .withFileSizeInBytes(stat.getLen())
+            .withMetrics(metrics)
+            .withPartitionPath(partitionKey)
+            .build();
+  }
 
-          }).collect(Collectors.toList());
-    } catch (IOException e) {
-      throw new RuntimeException("Unable to list files in partition: " + partitionUri, e);
-    }
+  private static ExecutorService migrationService(int concurrentDeletes) {
```

Comment:
@RussellSpitzer, does this get shut down?

Reply:
Only through GC, like the previous procedures https://github.com/apache/iceberg/blob/7fc8f9a6c8fe9dacb2c19fdfa76dd8c592c47c24/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/ExpireSnapshotsProcedure.java#L107

And

https://github.com/apache/iceberg/blob/7249d67c5b5215bd04d51d446de12934d2efcbae/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RemoveOrphanFilesProcedure.java#L118

We could close it out explicitly though in all of these uses

Reply:
We probably should close explicitly rather than waiting. But good to know that they're closed when they get garbage collected.

Reply:
Looking at this I think we should push this parameter through as an executor service the whole way so actions can reuse a Service (like in ExpireSnapshots or RemoveOrphans). I think i'll do one PR for the other procedures then work with @kingeasternsun to modify https://github.com/apache/iceberg/pull/3973/files to fit.

My basic thought process is 

Actions with parallelism get a wtihExecutorService() method
Procedures who use that follow a 
Create Service
Execute Action
Close Service pattern.

So we extract the service out of the utility class and into the procedure itself

Reply:
I was thwarted in my plans for this particular method, the issue is we call this on the executors and not on the driver so we can't pass through an executor service, at least not in a sensible way so I think we just need to close it in the utility method ... not ideal

Reply:
#4125

Reply:
> #4125

Thanks